server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -206,18 +204,35 @@ private static MapperParsingException wrapInMapperParsingException(SourceToParse
     }
 
     /** Creates a Mapping containing any dynamically added fields, or returns null if there were no dynamic mappings. */
-    static Mapping createDynamicUpdate(Mapping mapping, DocumentMapper docMapper, List<Mapper> dynamicMappers) {
-        if (dynamicMappers.isEmpty()) {
+    static Mapping createDynamicUpdate(Mapping mapping,
+                                       DocumentMapper docMapper,
+                                       List<Mapper> dynamicMappers,
+                                       List<RuntimeFieldType> dynamicRuntimeFields) {
+        if (dynamicMappers.isEmpty() && dynamicRuntimeFields.isEmpty()) {
             return null;
         }
+        RootObjectMapper root;
+        if (dynamicMappers.isEmpty() == false) {
+            root = createDynamicUpdate(mapping.root, docMapper, dynamicMappers);
+        } else {
+            root = (RootObjectMapper)mapping.root.mappingUpdate(null);
```

Comment:
this is pretty horrible. I am not sure how to make it better yet. We are introducing one reason to update the root without having to add a mapper to it.

Reply:
Passing `null` to make a copy is a bit sad, yeah. Maybe we could have a `copy` method?

Reply:
I followed this approach and things do look a bit better, thanks!